[login] *BREAKING CHANGE* login server rewrite

zach2good opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Fixes https://github.com/LandSandBoat/server/issues/2541

Hold onto your butts, we're getting rid of manual management of sockets and replacing it with [ASIO](https://think-async.com/Asio/) (this fd leaking is the reason for the issue above, which should be fixed by this rewrite). 

A rewrite also means we can get rid of all of the old manual and arcane c-style socket and packet handling. Hopefully this leads to code people aren't afraid of looking at. 

***This is a breaking change that needs a new xiloader distributed to clients***
https://github.com/LandSandBoat/xiloader/pull/7 MUST be merged and this new xiloader MUST be released.

Future **TODO**

- Flag rename of characters ?
- Salt and hash passwords (how do we migrate this from existing accounts?)
- TOTP
- Re-hook-up the message server process (low priority, message handling does not do anything for login yet)

**DONE**

- Clean up into `application.h`
- Bring proper `spdlog` logging back in
- Do the sign up and login process
- TLS integration with OpenSSL
- Self signed certificate auto generation
- Load existing certs (if you so decide to want to pay for a cert)
- Support displaying characters in different sizes (height)
- Add setting for disabling character creation (generally for external website character creation)
-  https://github.com/LandSandBoat/server/pull/3520

Documentation will need to be updated on using existing certs.

## Steps to test these changes
Use xiloader from this PR https://github.com/LandSandBoat/xiloader/pull/7
- Log in as normal
- Create accounts as normal

zach2good committed:
[login] Begin login rewrite using asio

WinterSolstice8 commented:
maintainable code, for one.

zach2good commented:
Updated description

kaincenteno commented:
Awesome! Thanks for clarification. =D

The only thing xiweb has had issues that might be related is that if a char doesn't clean shutdown. It will still show up in the chars online dB and the char will be unable to log in to ximap until the login and log off from xiconnect

zach2good commented:
Is there an issue for that?

kaincenteno commented:
> Is there an issue for that?

thought there was, my bad xD. Now there is https://github.com/LandSandBoat/server/issues/3054 :D 

I noticed this because we use the accounts_session on xiweb site to know who is currently online. (Currently the site queries to the xidb are broken as Im waiting for a router to arrive [dont wanna hardcode IPs of the game server])

zach2good committed:
start session tracking

zach2good committed:
Add AUTH_COMPONENTS

zach2good committed:
Update compiler flags for WolfSSL

WinterSolstice8 committed:
[cmake] Update ext libs to disable Wolfssl examples, crypt tests

* Update git tag to latest wolfssl release

WinterSolstice8 committed:
[login] Add basic authentication

WinterSolstice8 committed:
[login] Move expansions & feature display to login settings

WinterSolstice8 committed:
[headlessxi] Change headlessxi to default to character slot 0

* Remove printing of SessionHash, as its not human readable
* Add logging if character name could not be located

WinterSolstice8 committed:
[headlessxi] Properly pack server IP into 0xA1 packets

* Add optional debug packets parameter to hxi constructor

WinterSolstice8 committed:
[CI] add additional table logging to login & startup checks

WinterSolstice8 committed:
[CI] [login] Update login & headlessxi for better CI

* This changes debug flags on headlessxi client to be less spammy
* Adds needed database entries for headlessxi login
* Moves some error logging to trace in login, it was needless noise from errors that are unavoidable and now allows CI to pass
* Adds a "mysqlcommand" helper to headlessxi db building during full linux starup checks. This helps facilitate better debugging if things go wrong.

WinterSolstice8 commented:
Companion PR: https://github.com/LandSandBoat/xiloader/pull/7

WinterSolstice8 committed:
[login] Change MAC address retrieval from xiloader to reserved space

* MAC address is no longer being sent from xiloader

WinterSolstice8 committed:
[login] Increase rlimit for open files to 10240

Xaver-DaRed committed:
Add migration script for original accid

WinterSolstice8 committed:
[login] Support character IDs above 65535

WinterSolstice8 committed:
[core] Remove remnants of unused message server in login

WinterSolstice8 committed:
[login] Avoid perma-lockout if you never actually got a map session

WinterSolstice8 committed:
[login] Send version update required error through view_session

* Add FFXI client info "Setting up connection" to 0x26 case in view session.

WinterSolstice8 committed:
[login] Add TLS support for authentication socket using OpenSSL

WinterSolstice8 commented:
review away

zach2good commented on .github/workflows/build.yml:
```diff
@@ -435,51 +435,90 @@ jobs:
           sleep 300s
 
+          # define bash helper function to help diagnose any sql errors
+          function mysqlcmd() { mysql xidb -h 127.0.0.1 -uroot -proot --verbose -e "$@"; }
```
👍

zach2good commented on cmake/CompilerWarnings.cmake:
```diff
@@ -19,6 +19,9 @@ function(set_project_warnings project_name)
       /wd4459 # declaration of 'var' hides global declaration
 
+      # Very common in WolfSSL
+      /wd4820 # 'n' bytes padding added after data member 'member'
```
Can you check to see if this is needed anymore if we're dumping WOlfSSL?

WinterSolstice8 replied:
Removed

zach2good commented on cmake/FindOpenSSLlibcrypto.cmake:
```diff
@@ -0,0 +1,29 @@
+        /usr/lib/
+        /usr/local/
+        /usr/local/bin/
+		/usr/local/opt/openssl/lib/ # OSX brew install location
```
Tabs

WinterSolstice8 replied:
Fixed

zach2good commented on cmake/FindOpenSSLlibssl.cmake:
```diff
@@ -0,0 +1,29 @@
+        /usr/lib/
+        /usr/local/
+        /usr/local/bin/
+		/usr/local/opt/openssl/lib/ # OSX brew install location
```
Tabs

WinterSolstice8 replied:
Fixed

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+    // clang-format on
+}
+
+class handler_session
```
This is all going to need to be split out into a cpp file 💀

WinterSolstice8 replied:
I split it out into .h files, is that enough? please review

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+    return fmt::format("{}", address);
+}
+
+uint32 str2ip(const char* ip_str)
```
All of these "global" helper functions used just in here should be in an anonymous namespace in the cpp file for connect_server

WinterSolstice8 replied:
Done, but I moved it to a helper file because many classes access those "global" functions

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+
+    packet[8] = 0x04; // result
+
+    packet[28] = 0x10; // Unknown
```
Do we have any idea what this is yet? Worth asking atom0s?

WinterSolstice8 replied:
Hmm. I suppose it might be worth asking about it. This is just "what was there" in the previous iteration.

WinterSolstice8 replied:
https://github.com/atom0s/XiPackets/blob/main/lobby/S2C_0x0004_ResponseError.md -> `// This field is never referenced within the client. It was always observed to be 0x10, but the actual value or its purpose is unknown.`
Turns out my comment was apparently correct. I've added a comment and a reference.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+{
+    uint16 mask = EXPANSION_DISPLAY::BASE_GAME;
+
+    if (settings::get<bool>("login.RISE_OF_ZILART"))
```
All of these ifs could be replaced with that pattern we use in lua:

table = {{ "login.CHAINS_OF_PROMATHIA", EXPANSION_DISPLAY::CHAINS_OF_PROMATHIA}, { ...}, ... }

for entry in table:
    if setting_get(first):
        mask |= second


etc.

WinterSolstice8 replied:
Done

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+{
+    uint16 mask = 0;
+
+    if (settings::get<bool>("login.SECURE_TOKEN"))
```
Same loop idea here

WinterSolstice8 replied:
Done

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        return -1;
+    }
+
+    Query = "INSERT INTO char_look(charid,face,race,size) VALUES(%u,%u,%u,%u);";
+
+    if (sql->Query(Query, charid, createchar->m_look.face, createchar->m_look.race, createchar->m_look.size) == SQL_ERROR)
```
We don't need to be reuding the Query object everywhere  here:

```
if (sql->Query("The query string here",
    charid, createchar->m_look.face, createchar->m_look.race, createchar->m_look.size) == SQL_ERROR)
```

WinterSolstice8 replied:
I moved most of them that made sense. Some of the larger ones were formatted nicely and I left them alone.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+{
+    auto sql = std::make_unique<SqlConnection>();
+    // Seed the random number generator.
+    srand(clock());
```
Do we need this? Presumably a little later we're going to use xirand for something, so that fine?

WinterSolstice8 replied:
we're using xirand now

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+            return;
+        }
+
+        /*auto componentFlags = ref<uint64>(data_, 1);
```
Dead code to be removed?

WinterSolstice8 replied:
Essentially a "demo" on how to use the feature flags that are currently not used because there's no newer versions of xiloader yet that use any.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        std::memcpy(usernameBuffer, data_ + 0x09, 16);
+        std::memcpy(passwordBuffer, data_ + 0x19, 16);
+
+        // std::string reservedData(data_ + 0x40, 17);
```
Dead code?

WinterSolstice8 replied:
Replaced with a comment

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        }*/
+
+        char usernameBuffer[17] = {};
+        char passwordBuffer[17] = {};
```
While we're doing a breaking change, should/could we increase the max password length to 32? and how hard/clearly do we enforce it here and on xiloader?

These lengths are used a few times, extract them out somewhere?

WinterSolstice8 replied:
password is 32 characters now

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        auto sql = std::make_unique<SqlConnection>();
+
+        char escaped_name[16 * 2 + 1] = {};
+        char escaped_pass[32 * 2 + 1] = {};
```
Just as I say that, theres a 16/32 discrepancy here?

WinterSolstice8 replied:
password is 32 characters now

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                        {
+                            while (sql->NextRow() == SQL_SUCCESS)
+                            {
+                                /*uint32 charid = sql->GetUIntData(0);
```
dead code?

WinterSolstice8 replied:
Dead but intentionally left in if we ever want to restore `MSG_LOGIN` ZMQ. I can remove in the interim if wanted.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        do_read();
+    }
+
+    void handle_error(std::error_code ec, std::shared_ptr<handler_session> self) override
```
Should this log?

WinterSolstice8 replied:
Logging errors here is very noisy. I've disabled it due to the fact that any early disconnection logs. It would be quite spammy with little info of use. I've added a comment to reflect as such.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+
+    for (size_t y = 0; y < size; y++)
+    {
+        std::memset(buffer, 0, sizeof(buffer));               // TODO: Replace these three lines with std::format when/if we move to C++ 20.
```
God I hate this, can you try and use fmt lib/spdlog to kill this?

WinterSolstice8 replied:
Killed as requested

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+    view_session(asio::ssl::stream<asio::ip::tcp::socket> socket)
+    : handler_session(std::move(socket))
+    {
+        DebugSockets("view_session");
```
In all of these debug prints, is it possible to also print out some information about the incoming connection/target socket?

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+
+        if (sessionHash == "")
+        {
+            ShowWarning("Session requested without valid sessionHash!");
```
For all of these warnings, we probably also want to log the incorect/abusive ip/account

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                }
+                else
+                {
+                    // TODO: what's the error code to client here?
```
Can we handle this?

WinterSolstice8 replied:
Couldn't produce this error on retail even if I wanted to. I have no tooling for this, but I could just throw an error at them anyway.

WinterSolstice8 replied:
This was replaced with closing the socket. It will produce an error.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+            break;
+            case 0x1F: // 31: "Acquiring Player Data"
+            {
+                // ShowInfo("Data send triggered from view");
```
Dead code?

WinterSolstice8 replied:
Removed

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                }
+                else
+                {
+                    generateErrorMessage(data_, 332); // "Could not connect to lobby server.\nPlease check this title's news for announcements."
```
Can we get all the error messages made up into an enum?

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+
+                data_[8] = 0x23; // result
+
+                data_[28] = 0x20; // unknown
+                data_[32] = 0x64; // unknown
```
Ask atom0s if he knows what these are?

WinterSolstice8 replied:
Replaced completely, no longer an issue.

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+        do_read();
+    }
+
+    void handle_error(std::error_code ec, std::shared_ptr<handler_session> self) override
```
Should this log?

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                    unsigned char CharList[2500];
+                    std::memset(CharList, 0, sizeof(CharList));
+
+                    // Store the reserved numbers.
```
Do we know what these are?

WinterSolstice8 replied:
Replaced completely, no longer an issue

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                    }
+                    else
+                    {
+                        // TODO: throw error to client here?
```
Should this log?

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+            break;
+            case 0xA2: // 162 : "Notifying lobby of current selections" pt 2
+            {
+                LOBBY_A2_RESERVEPACKET(ReservePacket);
```
All of these macro things: LOBBY_A2_RESERVEPACKET
Can we refactor them out to be functions that return an object instead of dumping something into the current scope?

WinterSolstice8 replied:
Refactored out into oblivion, now is a proper class provided by atom0s

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,2257 @@
+                }
+            }
+
+            if (Application::IsRunning())
```
Given how long this file is, consider how you might be able to break it up into smaller files (given that there still needs to be a .h/.cpp cleanup already)

WinterSolstice8 replied:
Broke out everything that made sense into .h files

zach2good commented on src/login/login.cpp:
```diff
@@ -66,7 +66,15 @@ struct epoll_event login_lobbyviewEpollEvent = { EPOLLIN };
 #include "login_conf.h"
+#include "message_server.h"
 
+/* wolfSSL */
```
All of this old stuff can die. Do the current login/lobby files need to exist anymore?

WinterSolstice8 replied:
File deleted

zach2good commented on src/login/login.cpp:
```diff
@@ -196,6 +204,12 @@ int32 do_init(int32 argc, char** argv)
     ShowInfo("The login-server is ready to work!");
     ShowInfo("=======================================================================");
 
+    WOLFSSL_CTX* ctx = nullptr;
```
RIP

WinterSolstice8 replied:
File deleted

zach2good reviewed (commented):
Looking good! (I can't approve/reject because I opened this PR originally)

In summary:
- Make sure we log for all errors/suspicious activity, with details about whos doing it and whats going on
- Turn all FFXI client error packets into an enum
- Remove all traces of WolfSSL and earlier work
- If we can, kill as much of the old login server code as we can
- Split connect_server out into h/cpp, and possibly other utils
- Break in-file utils out into an anon namespace
- Kill any dead commented out code you can find
- Any magic numbers, cross-reference with atom0s to see if he can help give them names at least

Top top work!

WinterSolstice8 committed:
[login] Cleanup, split out connect_server.h, delete old files.

WinterSolstice8 commented:
Still need to address the security warning logs, currently using CI to tell me if things still work.

WinterSolstice8 commented:
Alright, go ahead and give it a second pass.

WinterSolstice8 committed:
[login] Move classes implementation out to .cpp files

zach2good commented on src/login/auth_session.cpp:
```diff
@@ -0,0 +1,389 @@
+        }
+    });
+    // clang-format on
+}
```
Newline at end of file

WinterSolstice8 replied:
Done

zach2good commented on src/login/connect_server.h:
```diff
@@ -0,0 +1,210 @@
+#include "session.h"
+#include "view_session.h"
+
+class ConnectServer final : public Application
```
This file also needs the cpp treatment

WinterSolstice8 replied:
Done

zach2good commented on src/login/login_errors.h:
```diff
@@ -0,0 +1,35 @@
+        GAMES_DATA_HAS_BEEN_UPDATED             = 331, // "The games data has been updated."
+        COULD_NOT_CONNECT_TO_LOBBY_SERVER       = 332, // "Could not connect to lobby server.\nPlease check this title's news for announcements."
+    };
+}
```
Newline at end of file

WinterSolstice8 replied:
Done

zach2good commented on src/login/login_errors.h:
```diff
@@ -0,0 +1,35 @@
+    {
+        CHARACTER_NAME_UNAVAILABLE              = 313, // "The character name you entered is unavailable.\nPlease choose another name." // This message is displayed in Japanese, needs fixing.
+        FAILED_TO_REGISTER_WITH_THE_NAME_SERVER = 314, // "Failed to register with the name server."
+        UNABLE_TO_CONNECT_TO_WORLD_SERVER       = 305, // "Unable to connect to the world server. Specified operation failed."
```
Can these be sorted by value?

WinterSolstice8 replied:
Done

zach2good commented on src/login/login_errors.h:
```diff
@@ -0,0 +1,35 @@
+{
+    enum errorCode : uint16
+    {
+        CHARACTER_NAME_UNAVAILABLE              = 313, // "The character name you entered is unavailable.\nPlease choose another name." // This message is displayed in Japanese, needs fixing.
```
Can you make this comment a TODO, and maybe a little bit more prominent - on the line above?

WinterSolstice8 replied:
Done

zach2good reviewed (commented):
(I can't approve because I started this PR, but if I could, I would)

In terms of actual content, ✅ { Approved }!, ship it!

Just a couple of style things I'd like fixed before merge:
- Not all files have the license headers
- Some files are missing a newline at the end of the file (I thought we had CI checks for this? Maybe not? Maybe its only checking for Win newlines and not Unix?)
- There are some .h files that have function implementations without an associated .cpp file. Can we get those moved into .cpp files so that changes only affect that compilation unit? If its just a data type or an interface, then just a header on its own is fine - but if there are implementation bits, to a cpp file please

After merge we can go through and make all the TODOs into GH issues with a login tag. Once we kill all TODOs, theoretically login server is feature complete and nobody ever has to touch it again?

Well done, you hero!

WinterSolstice8 committed:
[login] Further split-out into .cpp files

Add new setting to disable character creation

Fix password changing to use new offset and lengths

WinterSolstice8 committed:
Add OpenSSL 3.1.1 binaries for x86_64 and 32 bit Windows
